### PR TITLE
add remat tutorial docs

### PR DESCRIPTION
**Pull request by mattjj**
We've never had a tutorial on doing autodiff checkpointing (aka rematerialization) with `jax.checkpoint` / `jax.remat`! And it's subtle enough to need an explanation.

Also, this is the first real documentation for _custom AD checkpointing policies_, which can be a key ingredient in getting [the highest model FLOP utilization for Transformer LLMs](https://twitter.com/zhangqiaorjc/status/1524822247860424704). We've had the feature for more than a year, but kept it secret by not documenting it :P

TODO:
- [x] update in light of https://github.com/google/jax/pull/13859

### CODE REVIEW

**Review by froystig — CHANGES_REQUESTED** [review]
There's something off with how this is [rendering](https://jax--13756.org.readthedocs.build/en/13756/) at the moment.

I'm seeing two distinct bullets, under the "Advanced tutorials" section, like so:

* [Utilities](https://jax--13756.org.readthedocs.build/en/13756/notebooks/autodiff_remat.html)
* [Control autodiff’s saved values with jax.checkpoint (aka jax.remat)](https://jax--13756.org.readthedocs.build/en/13756/notebooks/autodiff_remat.html#control-autodiff-s-saved-values-with-jax-checkpoint-aka-jax-remat)

Still need to review, highlighting this for now.

File: docs/notebooks/autodiff_remat.md [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.1
+kernelspec:
+  display_name: Python 3
+  name: python3
+---
+
```
Comment (jakevdp) [inline docs/notebooks/autodiff_remat.md:13]: You need a top-level markdown heading at the top of this notebook to be the TOC title... otherwise sphinx does weird things with the table of contents.

Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:13]: Bleh. I wanted to have the "utilities" section collapsed at the top _before_ the title. But I guess that's not possible...

File: docs/notebooks/autodiff_remat.md [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+This scan-over-layers version reduces compile times, but by foiling some XLA optimizations it can lead to inefficient computation of gradients. To mitigate the issue, we would use `jax.checkpoint` on the scanned function:
+
++++ {"id": "iHVVNVdO66Dv"}
+
+```python
+from functools import partial
+
+@partial(jax.checkpoint,
+         policy=jax.checkpoint_policies.checkpoint_dots_with_no_batch_dims)
```
Comment (LenaMartens) [inline docs/notebooks/autodiff_remat.md:539]: out of curiosity, is this the trick to get to "getting [the highest model FLOP utilization for Transformer LLMs](https://twitter.com/zhangqiaorjc/status/1524822247860424704)" you refer to in the PR description?

**Review by LenaMartens — APPROVED** [review]
Great notebook, I think this helps a lot with building intuition around `remat`! I'd merge it ASAP 😁

File: docs/notebooks/autodiff_remat.md [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+
+### Practical notes
+
++++ {"id": "LSADkBOCyX9X"}
+
+When differentiated functions are staged out to XLA for compilation, for example by applying `jax.jit` to a function which contains a `jax.grad` call, XLA will automatically optimize the computation, including decisions about when to compute or rematerialize values. As a result, **`jax.checkpoint` often isn't needed for differentiated functions under a `jax.jit`**. XLA will optimize things for you.
+
+One exception is when using staged-out control flow, like `jax.lax.scan`. Automatic compiler optimizations between multiple control flow primitives, e.g. between a forward-pass `scan` and the corresponding backward-pass `scan`, aren't aren't as thorough. As a result, it's often a good idea to use `jax.checkpoint` on the body function we pass to `jax.lax.scan`.
+
+For example, one common pattern in large Transformer models is to express the architecture as a `jax.lax.scan` over layers so as to reduce compilation times. That is, using a simple fully-connected network as an analogy, instead of writing something like this:
```
Comment (froystig) [inline docs/notebooks/autodiff_remat.md:490]:
```suggestion
For example, one common pattern in large [Transformer](https://arxiv.org/abs/1706.03762) models is to express the architecture as a `jax.lax.scan` over layers so as to reduce compilation times. To use a simple fully-connected network as an analogy, instead of writing something like this:
```
or, alternatively, link to the [Wikipedia article](https://en.wikipedia.org/wiki/Transformer_(machine_learning_model))?

File: docs/notebooks/autodiff_remat.md [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+
+def net(params: ParamsList, x: jnp.ndarray):
+  for W, b in params:
+    x = jnp.maximum(jnp.dot(x, W) + b, 0.)
+  return x
+```
+
++++ {"id": "38xDcRwW518P"}
+
+We would instead use `jax.lax.scan`:
```
Comment (froystig) [inline docs/notebooks/autodiff_remat.md:508]:
```suggestion
We would instead iterate the layer application with `jax.lax.scan`:
```

File: docs/notebooks/autodiff_remat.md [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+  return out, None
+
+def net(all_weights, all_biases, x):
+  x, _ = jax.lax.scan(layer, x, (all_weights, all_biases))
+  return x
+```
+
++++ {"id": "A-NbCn8A6TFK"}
+
+This scan-over-layers version reduces compile times, but by foiling some XLA optimizations it can lead to inefficient computation of gradients. To mitigate the issue, we would use `jax.checkpoint` on the scanned function:
```
Comment (froystig) [inline docs/notebooks/autodiff_remat.md:531]:
```suggestion
This scan-over-layers version reduces compile times, but by foiling some compiler optimizations it can lead to inefficient computation of gradients. To mitigate the issue, we would use `jax.checkpoint` on the scanned function:
```

**Review by froystig — APPROVED** [review]
Just some suggestions, up to you whether to accept/reject/amend each one!

File: docs/notebooks/autodiff_remat.md [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+
+# Inspect the 'residual' values to be saved on the forward pass
+# if we were to evaluate `jax.grad(f)(W1, W2, W3, x)`
+from jax.ad_checkpoint import print_saved_residuals
+jax.ad_checkpoint.print_saved_residuals(f, W1, W2, W3, x)
+```
+
++++ {"id": "97vvWfI-fSSF"}
+
+By applying `jax.checkpoint` to sub-functions, as a decorator or at specific application sites, we force JAX not to save any of that sub-function's residuals. Instead, only the inputs of a `jax.checkpoint`-decorated function can be saved, and any residuals are computed on the backward pass from those as needed:
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:57]: Actually I think they may not actually be saved if they're not needed on the backward pass. In addition to XLA DCE, we actually do our own DCE to that effect [here](https://github.com/google/jax/blob/64145393b2c65bb011ca10d41314a8a06c025cb7/jax/_src/ad_checkpoint.py#L489) and [here](https://github.com/google/jax/blob/64145393b2c65bb011ca10d41314a8a06c025cb7/jax/_src/ad_checkpoint.py#L478).

File: docs/notebooks/autodiff_remat.md [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+def f2(W1, W2, W3, x):
+  x = jax.checkpoint(g)(W1, x)
+  x = jax.checkpoint(g)(W2, x)
+  x = jax.checkpoint(g)(W3, x)
+  return x
+
+jax.ad_checkpoint.print_saved_residuals(f2, W1, W2, W3, x)
+```
+
++++ {"id": "CyRR3mTpjRtl"}
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:69]: It's because those `sin` values are inputs to subsequent `jax.checkpoint(g)` applications. I'll add a comment!

File: docs/notebooks/autodiff_remat.md [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+
+To control which values are saveable without having to edit the definition of the function to be differentiated, you can use a rematerialization _policy_, for example to save only results whose computation is likely FLOP-bound:
+
+```{code-cell}
+f3 = jax.checkpoint(f, policy=jax.checkpoint_policies.checkpoint_dots_with_no_batch_dims)
+jax.ad_checkpoint.print_saved_residuals(f3, W1, W2, W3, x)
+```
+
++++ {"id": "9fe6W0YxlfKa"}
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:78]: Thanks so much for flagging this. I agree! I'll add aliases and change these docs to refer to `dots_saveable` etc.

File: docs/notebooks/autodiff_remat.md [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+# no use of jax.checkpoint:
+print_fwd_bwd(f, W1, W2, W3, x)
+```
+
+```{code-cell}
+# using jax.checkpoint with policy=jax.checkpoint_policies.checkpoint_dots_with_no_batch_dims:
+print_fwd_bwd(f3, W1, W2, W3, x)
+```
+
++++ {"id": "UsvnQJYomcub"}
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:148]: Dark mode is hereby banned!

File: docs/notebooks/autodiff_remat.md [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+    x_bar, = g_vjp(y_bar)
+    return x_bar
+  return z, f_bwd2
+```
+
++++ {"id": "JYMw6oxtp6SH"}
+
+In words, this alternative implementation doesn't compute `g_vjp`, or the residual values in its closure, on the forward pass. Instead it only computes them in the backward pass `f_bwd2`. That means `f_vjp_checkpoint` requires less memory: if `g` and `h` each required similar amounts of memory for their residuals, each much larger than `x`, then the function produced by `f_vjp_checkpoint(x)` requires half the memory as that of `f_vjp(x)`!
+
+The cost we pay is redundant work: in `f_bwd2` we must re-evaluate `g(x)` just to discard its value (in the underscore variable on the line `_, g_vjp = jax.vjp(g, x)`).
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:223]: Good idea, done!

File: docs/notebooks/autodiff_remat.md [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+  _, h_vjp = jax.vjp(h)(y)  # step 2
+  y_bar, = h_vjp(1.0)       # step 3
+  _, g_vjp = jax.vjp(g, x)  # step 4
+  x_bar, = g_vjp(y_bar)     # step 5
+  return x_bar
+```
+
++++ {"id": "tfssPyb2sQgw"}
+
+In general, `jax.checkpoint(foo)` is a new function which has the same input-ouptut behavior as `foo`, but behaves differently under autodiff, particularly under `jax.linearize` and `jax.vjp` (and their wrappers, like `jax.grad`) but not `jax.jvp`. When differentiated, only the input to a `jax.checkpoint`-differentiated function is stored on the forward pass; on the backward pass, residuals (i.e. intermediates from `foo` and its Jacobian coefficient values needed for the backward pass) are recomputed.
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:259]: Done!

File: docs/notebooks/autodiff_remat.md (hunk 2 above)
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:539]: Yeah!

File: docs/notebooks/autodiff_remat.md [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+import jax.numpy as jnp
+```
+
++++ {"id": "qaIsQSh1XoKF"}
+
+### TL;DR
+
+Use the `jax.checkpoint` decorator (aliased as `jax.remat`) with `jax.grad` to control which intermediates are saved on the forward pass versus recomputed on the backward pass, trading off memory and FLOPs.
+
+**Don't miss the [Practical notes](#practical-notes) for a discussion about how `jax.checkpoint` interacts with `jax.jit`.**
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:29]: Done!

File: docs/notebooks/autodiff_remat.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:57]: Excellent, thank you!

File: docs/notebooks/autodiff_remat.md [hunk 13]
```diff
@@ -0,0 +1,548 @@
...
+  x = jax.checkpoint(g)(W2, x)
+  x = jax.checkpoint(g)(W3, x)
+  return x
+
+jax.ad_checkpoint.print_saved_residuals(f2, W1, W2, W3, x)
+```
+
++++ {"id": "CyRR3mTpjRtl"}
+
+To control which values are saveable without having to edit the definition of the function to be differentiated, you can use a rematerialization _policy_, for example to save only results whose computation is likely FLOP-bound:
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:71]: Thanks! Tweaked a little to underscore that it's batch-less dots which are often FLOP-bound.

File: docs/notebooks/autodiff_remat.md [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+
++++ {"id": "qaIsQSh1XoKF"}
+
+### TL;DR
+
+Use the `jax.checkpoint` decorator (aliased as `jax.remat`) with `jax.grad` to control which intermediates are saved on the forward pass versus recomputed on the backward pass, trading off memory and FLOPs.
+
+**Don't miss the [Practical notes](#practical-notes) for a discussion about how `jax.checkpoint` interacts with `jax.jit`.**
+
+Without using `jax.checkpoint`, the forward pass of `jax.grad(f)(x)` saves the values of some intermediates, and computes and saves the values of Jacobian coefficients, for use on the backward pass. We call these saved values _residuals_:
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:31]: Done!

File: docs/notebooks/autodiff_remat.md [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+  return x_bar
+```
+
++++ {"id": "tfssPyb2sQgw"}
+
+In general, `jax.checkpoint(foo)` is a new function which has the same input-ouptut behavior as `foo`, but behaves differently under autodiff, particularly under `jax.linearize` and `jax.vjp` (and their wrappers, like `jax.grad`) but not `jax.jvp`. When differentiated, only the input to a `jax.checkpoint`-differentiated function is stored on the forward pass; on the backward pass, residuals (i.e. intermediates from `foo` and its Jacobian coefficient values needed for the backward pass) are recomputed.
+
++++ {"id": "HVvS_S5zsVZ-"}
+
+Notice that with `f = lambda x: h(g(x))` is the funciton we want to differentiate, i.e. if we want to apply `jax.grad(f)`, we don't get any memory savings by applying `jax.checkpoint` to `f` itself. That's because evaluating `jax.grad(jax.checkpoint(f))(x)` would lead to a computation like:
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:263]: Done!

File: docs/notebooks/autodiff_remat.md [hunk 16]
```diff
@@ -0,0 +1,548 @@
...
+
++++ {"id": "JYMw6oxtp6SH"}
+
+In words, this alternative implementation doesn't compute `g_vjp`, or the residual values in its closure, on the forward pass. Instead it only computes them in the backward pass `f_bwd2`. That means `f_vjp_checkpoint` requires less memory: if `g` and `h` each required similar amounts of memory for their residuals, each much larger than `x`, then the function produced by `f_vjp_checkpoint(x)` requires half the memory as that of `f_vjp(x)`!
+
+The cost we pay is redundant work: in `f_bwd2` we must re-evaluate `g(x)` just to discard its value (in the underscore variable on the line `_, g_vjp = jax.vjp(g, x)`).
+
++++ {"id": "LqTrjPoGqrK7"}
+
+We can get this autodiff behavior without having to write VJP functions directly by instead using `jax.checkpoint` in an alternative definition of `f`:
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:227]: Done!

File: docs/notebooks/autodiff_remat.md [hunk 17]
```diff
@@ -0,0 +1,548 @@
...
+
++++ {"id": "wvSm1yu0yUtd"}
+
+### Practical notes
+
++++ {"id": "LSADkBOCyX9X"}
+
+When differentiated functions are staged out to XLA for compilation, for example by applying `jax.jit` to a function which contains a `jax.grad` call, XLA will automatically optimize the computation, including decisions about when to compute or rematerialize values. As a result, **`jax.checkpoint` often isn't needed for differentiated functions under a `jax.jit`**. XLA will optimize things for you.
+
+One exception is when using staged-out control flow, like `jax.lax.scan`. Automatic compiler optimizations between multiple control flow primitives, e.g. between a forward-pass `scan` and the corresponding backward-pass `scan`, aren't aren't as thorough. As a result, it's often a good idea to use `jax.checkpoint` on the body function we pass to `jax.lax.scan`.
```
[the remark this answers is not in the document]
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:488]: Done!

File: docs/notebooks/autodiff_remat.md (hunk 3 above)
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:490]: Done!

File: docs/notebooks/autodiff_remat.md (hunk 4 above)
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:508]: Done!

File: docs/notebooks/autodiff_remat.md (hunk 5 above)
Comment (mattjj) [reply docs/notebooks/autodiff_remat.md:531]: Done!
